forem/forem
# Spike: Use Postgres full text search

## What type of PR is this? (check all applicable)

- [X] Spike

## Description

The purpose of this PR is to spike out using Postgres full-text search (FTS) instead of Elasticsearch for our search needs. This is a rough first cut I spent less than 2h on so far, the idea was mainly to illustrate how Postgres FTS works. If we decide to go ahead with this we can work on a proper RFC.

### Notes

* We're using the [pg_search](https://github.com/Casecommons/pg_search) for most of the heavy lifting. I've used this gem extensively in the past, but never in multi-search mode only with search scopes (see [Multi-search vs. search scope](https://github.com/Casecommons/pg_search#multi-search-vs-search-scopes)).
* I've only added search to two models (`Article` and `PodcastEpisode`) for now and only for two fields on each.
* I added a service for index management (`Search::Indices`) which is quite convenient for quickly doing things from a Rails console or similar. Note that the `pg_search` gem also comes with [Rake tasks](https://github.com/Casecommons/pg_search#rebuilding-search-documents-for-a-given-class) for this purpose.
* I enabled two Postgres extensions ([unaccent](https://www.postgresql.org/docs/13/unaccent.html) and [pg_trgm](https://www.postgresql.org/docs/13/pgtrgm.html) because every search I ever built needed them sooner than later.
* I added an example search service (`Search::Multisearch`) and specs to show how things work.

### What's next?

* [X] Adding a GIN index on `pg_search_documents.content`.
* [ ] Add example for `pg_search_scope`
* [ ] Fine-tune search options ("or" searches, negation, prefix searches, trigram config, etc.)
* [ ] Adding one or two more models
* [ ] Adding more attributes to the different models
* [ ] Implement at least one example of custom indexing with `rebuild_pg_search_documents`
* [ ] You tell me 😃 

## Related Tickets & Documents

https://github.com/forem/InternalProjectPlanning/issues/381

## QA Instructions, Screenshots, Recordings

1. Check out this branch.
2. Play around with `Search::Indices` from a `rails console`, e.g.
    ```ruby
    Search::Indices.clear!(Article)
    Search::Indices.rebuild!(Article)
    ```
3. Play around with the search function itself, e.g.
    ```ruby
    Search::Multisearch.call("search term") # depending on your seed data
    ```
    Note that the search options aren't fine-tuned yet, so things like partial matches, etc. won't work.

### UI accessibility concerns?

No UI yet.

## Added tests?


## Added to documentation?

- [X] No documentation needed, only a spike for now
